services/appflowy-collaborate/src/collab/storage.rs

```diff
@@ -75,6 +89,103 @@ where
     }
   }
 
+  const PENDING_WRITE_BUF_CAPACITY: usize = 100;
+  async fn periodic_write_task(cache: CollabCache, mut reader: Receiver<PendingCollabWrite>) {
+    let mut buf = Vec::with_capacity(Self::PENDING_WRITE_BUF_CAPACITY);
+    loop {
+      let n = reader
+        .recv_many(&mut buf, Self::PENDING_WRITE_BUF_CAPACITY)
+        .await;
+      if n == 0 {
+        break;
+      }
+      let pending = buf.drain(..n);
+      if let Err(e) = Self::persist(&cache, pending).await {
+        tracing::error!("failed to persist {} collabs: {}", n, e);
+      }
+    }
+  }
+
+  async fn persist(
+    cache: &CollabCache,
+    records: impl ExactSizeIterator<Item = PendingCollabWrite>,
+  ) -> Result<(), AppError> {
+    // Start a database transaction
+    let mut transaction = cache
+      .pg_pool()
+      .begin()
+      .await
+      .context("Failed to acquire transaction for writing pending collaboration data")
+      .map_err(AppError::from)?;
+
+    tracing::trace!("persisting {} collabs", records.len());
+    // Insert each record into the database within the transaction context
+    let mut action_description = String::new();
+    for (index, record) in records.into_iter().enumerate() {
+      let params = record.params;
+      action_description = format!("{}", params);
+      let savepoint_name = format!("sp_{}", index);
+
+      // using savepoint to rollback the transaction if the insert fails
+      sqlx::query(&format!("SAVEPOINT {}", savepoint_name))
+        .execute(transaction.deref_mut())
+        .await?;
+      if let Err(_err) = cache
+        .insert_encode_collab_to_disk(&record.workspace_id, &record.uid, params, &mut transaction)
+        .await
+      {
+        sqlx::query(&format!("ROLLBACK TO SAVEPOINT {}", savepoint_name))
+          .execute(transaction.deref_mut())
+          .await?;
+      }
+    }
+
+    // Commit the transaction to finalize all writes
+    match tokio::time::timeout(Duration::from_secs(10), transaction.commit()).await {
```

Comment:
Assuming we have PENDING_WRITE_BUF_CAPACITY set for each write, I’m not sure if 10 seconds is sufficient.

Reply:
This part is literally copy-pasted from current StorageQueue.

Reply:
I believe they are different. The StorageQueue processes 20 collabs at a time, whereas, in some cases, you process 100 collabs at once.
```rust
 let chunk_keys = consume_pending_write(&pending_write_set, 20, 5).await;
      if chunk_keys.is_empty() {
        continue;
      }

      for keys in chunk_keys {
        trace!(
          "start writing {} pending collaboration data to disk",
          keys.len()
        );
        let cloned_collab_cache = collab_cache.clone();
        let mut cloned_connection_manager = connection_manager.clone();
        let cloned_total_write_count = total_write_count.clone();
        let cloned_total_success_write_count = success_write_count.clone();

        if let Ok(metas) = get_pending_meta(&keys, &mut cloned_connection_manager).await {
          if metas.is_empty() {
            error!("the pending write keys is not empty, but metas is empty");
            return;
          }

          match retry_write_pending_to_disk(&cloned_collab_cache, metas).await {
            Ok(success_result) => {
              #[cfg(debug_assertions)]
              tracing::info!("success write pending: {:?}", keys,);

              trace!("{:?}", success_result);
              cloned_total_write_count.fetch_add(
                success_result.expected as i64,
                std::sync::atomic::Ordering::Relaxed,
              );
              cloned_total_success_write_count.fetch_add(
                success_result.success as i64,
                std::sync::atomic::Ordering::Relaxed,
              );
            },
            Err(err) => error!("{:?}", err),
          }
          // Remove pending metadata from Redis even if some records fail to write to disk after retries.
          // Records that fail repeatedly are considered potentially corrupt or invalid.
          let _ = remove_pending_meta(&keys, &mut cloned_connection_manager).await;
        }
      }
```

Reply:
I've set it to the same value as on main (20 collabs at a time).

Reply:
OK